src/transformStateWithClones.js

```diff
@@ -7,7 +7,48 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  const stateHistory = [];
+  let stateCopy = { ...state };
+
+  const addProperties = (extraData) => {
+    stateCopy = Object.assign(stateCopy, extraData);
+    stateHistory.push({ ...stateCopy });
+  };
+
+  const removeProperties = (keysToRemove) => {
+    for (const remove in keysToRemove) {
+      for (const keys in stateCopy) {
```

Comment:
There's no need to use a nested loop to remove properties. You can delete the property directly using 'delete stateCopy[keysToRemove[remove]]'. This will simplify the code and improve performance.